Python/Product/Analysis/Interpreter/Ast/AstAnalysisFunctionWalker.cs

```diff
@@ -0,0 +1,127 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.PythonTools.Analysis;
+using Microsoft.PythonTools.Infrastructure;
+using Microsoft.PythonTools.Parsing;
+using Microsoft.PythonTools.Parsing.Ast;
+
+namespace Microsoft.PythonTools.Interpreter.Ast {
+    class AstAnalysisFunctionWalker : PythonWalker {
+        private readonly FunctionDefinition _target;
+        private readonly NameLookupContext _scope;
+        private readonly List<IPythonType> _returnTypes;
+        private readonly AstPythonFunctionOverload _overload;
+
+        public AstAnalysisFunctionWalker(
+            NameLookupContext scope,
+            FunctionDefinition targetFunction
+        ) {
+            _scope = scope;
+            _target = targetFunction;
+            _returnTypes = new List<IPythonType>();
+            _overload = new AstPythonFunctionOverload(
+                AstPythonFunction.MakeParameters(_scope.Ast, _target),
+                _scope.GetLocOfName(_target, _target.NameExpression),
+                _returnTypes
+            );
+        }
+
+        public IList<IPythonType> ReturnTypes => _returnTypes;
+        public IPythonFunctionOverload Overload => _overload;
+
+        private void GetMethodType(FunctionDefinition node, out bool classmethod, out bool staticmethod) {
+            classmethod = false;
+            staticmethod = false;
+
+            if (node.IsLambda) {
+                staticmethod = true;
+                return;
+            }
+
+            var classmethodObj = _scope.Interpreter.GetBuiltinType(BuiltinTypeId.ClassMethod);
+            var staticmethodObj = _scope.Interpreter.GetBuiltinType(BuiltinTypeId.StaticMethod);
+            foreach (var d in (_target.Decorators?.Decorators).MaybeEnumerate()) {
```

Comment:
Without it you could end up trying to enumerate over `null`:
`_target.Decorators == null ? null : _target.Decorators.Decorators.MaybeEnumerate()`

With the parens, the `null` get passed to the extension method and we are guaranteed an empty enumerable:
`(_target.Decorators == null ? null : _target.Decorators.Decorators).MaybeEnumerate()`